### PR TITLE
fix: av1 codecs string

### PR DESCRIPTION
**Pull request by gnattu**
**Changes**

Browsers are very picky about this string. The profile level has to be padded to double digits otherwise the stream is going to be rejected and the playback would not continue.

This is confirmed to work with latest Chromium 122 and Firefox 124, but Chromium used to be more picky about this string so additional hacks to lying about the level to it may be needed.

**Issues**

Fixes jellyfin/jellyfin-web#5364

### CODE REVIEW

**Review by Shadowghost — APPROVED** [review]

File: Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs [hunk 1]
```diff
@@ -214,13 +214,15 @@ public static string GetAv1String(string? profile, int level, bool tierFlag, int
...
-        if (level <= 0
-            || level > 31)
+        if (level is <= 0 or > 31)
         {
             // Default to the maximum defined level 6.3
             level = 19;
         }
 
+        // Needed to pad it two double digits; otherwise, browsers will reject the stream.
+        var levelString = level < 10 ? $"0{level}" : $"level";
```
Comment (dmitrylyzo) [inline Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs:224]: Isn't `{}` missing?
```suggestion
        var levelString = level < 10 ? $"0{level}" : $"{level}";
```

File: Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs [hunk 2]
```diff
@@ -214,13 +214,15 @@ public static string GetAv1String(string? profile, int level, bool tierFlag, int
...
-        if (level <= 0
-            || level > 31)
+        if (level is <= 0 or > 31)
         {
             // Default to the maximum defined level 6.3
             level = 19;
         }
 
+        // Needed to pad it two double digits; otherwise, browsers will reject the stream.
+        var levelString = level < 10 ? $"0{level}" : $"{level}";
```
Comment (JPVenson) [inline Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs:224]: `level.ToString().PadLeft(2, '0')` ?

File: Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs [hunk 3]
```diff
@@ -230,7 +232,7 @@ public static string GetAv1String(string? profile, int level, bool tierFlag, int
         }
 
         result.Append('.')
-            .Append(level)
+            .Append(levelString)
```
Comment (Bond-009) [inline Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs:235]:
```suggestion
            .AppendFormat("{0:D2}", level)
```

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs (hunk 2 above)
Comment (JPVenson) [reply Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs:224]: also a good way

**Review by nyanmisaka — APPROVED** [review]
Good catch. The link in line 195 is outdated. Here’s the latest https://aomediacodec.github.io/av1-isobmff/#codecsparam

The level and bitdepth strings can be updated to use AppendFormat.

File: Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs (hunk 3 above)
Comment (gnattu) [reply Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs]: Done

File: Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs [hunk 4]
```diff
@@ -230,7 +229,8 @@ public static string GetAv1String(string? profile, int level, bool tierFlag, int
         }
 
         result.Append('.')
-            .Append(level)
+            // Needed to pad it double digits; otherwise, browsers will reject the stream.
+            .Append(CultureInfo.InvariantCulture, $"{level:D2}")
```
Comment (Bond-009) [inline Jellyfin.Api/Helpers/HlsCodecStringHelpers.cs:233]: Use AppendFormat instead of formatting a string and appending that

**Review by Bond-009 — CHANGES_REQUESTED** [review]

**Review by Bond-009 — APPROVED** [review]

**Pull request merged**